include/onnxruntime/core/framework/ortmemoryinfo.h

```diff
@@ -38,6 +38,20 @@ struct OrtMemoryInfo {
     return strcmp(name, other.name) < 0;
   }
 
+  static void HashCombine(size_t h, size_t& seed) {
+    seed ^= h + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+  }
+
+  // This is to make OrtMemoryInfo a valid key in hash tables
+  // we ignore device id
+  size_t Hash() const {
+    auto h = std::hash<int>()(alloc_type);
+    HashCombine(std::hash<int>()(mem_type), h);
+    HashCombine(std::hash<int>()(id), h);
+    HashCombine(std::hash<const char*>()(name), h);
+    return h;
+  }
+
```

Comment:
In most use cases the name doesn't matter. What does is the mem_type and device (which duplicates the id value). 

Should we have an alternative hash implementation that allows using that info in the hash so this knowledge is in one place? That could be used for SessionState::allocators_.

Reply:
I saw that alternate comparison function. After thinking a bit, I decided to make it equivalent of operator< here in the class. The way I see it:
1) We have special hash for allocators_ just like a special operator<
2) Or we change both of them here and use everywhere.

Reply:
operator< seems flawed. the OrtDevice is arguably the most important thing but is ignored by that implementation. 

it's fine to leave the allocator related hashing separate. it's a larger design discussion to rationalize the whole allocator setup and probably falls under that to decide what's best long term.

Reply:
As stated in this [link](http://en.cppreference.com/w/cpp/utility/hash):

> There is no specialization for C strings. std::hash produces a hash of the value of the pointer (the memory address), it does not examine the contents of any character array.

So is the std::hash<const char*> here an expected one?